MoreLinq/Experimental/Async/Merge.cs

```diff
@@ -0,0 +1,251 @@
+        /// order than their order in <paramref name="sources"/>.</para>
+        /// <para>
+        /// When disposed part of the way, there is a best-effort attempt to
+        /// cancel all iterations that are in flight. This requires that all
+        /// asynchronous streams in <paramref name="sources"/> properly
+        /// honour timely cancellation.</para>
+        /// </remarks>
+
+        public static IAsyncEnumerable<T> Merge<T>(this IEnumerable<IAsyncEnumerable<T>> sources,
+                                                   int maxConcurrent)
+        {
+            if (sources is null) throw new ArgumentNullException(nameof(sources));
+            if (maxConcurrent <= 0) throw new ArgumentOutOfRangeException(nameof(maxConcurrent));
+
+            return Async();
+
+            async IAsyncEnumerable<T> Async([EnumeratorCancellation]CancellationToken cancellationToken = default)
+            {
+                var thisCancellationTokenSource = new CancellationTokenSource();
+                var cancellationTokenSource = cancellationToken.CanBeCanceled
+                                            ? CancellationTokenSource.CreateLinkedTokenSource(thisCancellationTokenSource.Token, cancellationToken)
+                                            : thisCancellationTokenSource;
+                cancellationToken = cancellationTokenSource.Token;
+
+                var enumeratorList = new List<IAsyncEnumerator<T>>();
+
+                List<Task<(bool, IAsyncEnumerator<T>)>>? pendingTaskList = null;
+
+                try
+                {
+                    var enumerators =
+                        from source in sources
+                        select source.GetAsyncEnumerator(cancellationToken);
+
+                    enumeratorList.AddRange(enumerators);
+
+                    pendingTaskList = new List<Task<(bool, IAsyncEnumerator<T>)>>();
+
+                    const bool some = true;
+
+                    async ValueTask<(bool, T)> ReadAsync(IAsyncEnumerator<T> enumerator)
+                    {
+                        var task = enumerator.MoveNextAsync();
+
+                        if (task.IsCompleted)
+                        {
+                            if (await task.ConfigureAwait(false))
+                                return (some, enumerator.Current);
+
+                            await enumerator.DisposeAsync().ConfigureAwait(false);
+                            enumeratorList.Remove(enumerator);
+                        }
+                        else
+                        {
+                            pendingTaskList.Add(task.And(enumerator).AsTask());
+                        }
+
+                        return default;
+                    }
+
+                    while (enumeratorList.Count > 0)
+                    {
+                        while (pendingTaskList.Count is { } ptc
+                               && ptc < enumeratorList.Count
+                               && ptc < maxConcurrent)
```

Comment:
enhanced pattern matching?
```suggestion
                        while (pendingTaskList.Count is >= 0
                               && < enumeratorList.Count
                               && < maxConcurrent)
```

Reply:
Thanks, but did you even try that? I'm asking because it's illegal in C#? First of all `&&` would have to be `and` and you cannot use non-literals/constants with relational patterns so while comparing to `>= 0` is fine, you can't do `< enumeratorList.Count` or `< maxConcurrent`.

Reply:
Obviously not... 😉 Definitely I screwed up the `&&` vs `and`. However, I could have sworn I had used a variable value in the the range in some of my code last month. I did test it finally and you're right, each value must be constant. Alas... Excuse me while I go hang my head in shame. 🤦‍♂️😂